pd/src/testnet.rs

```diff
@@ -42,28 +45,101 @@ pub fn parse_validators(input: impl Read) -> Result<Vec<TestnetValidator>> {
     Ok(serde_json::from_reader(input)?)
 }
 
-/// Hardcoded Tendermint config template. Should produce tendermint config similar to
-/// https://github.com/tendermint/tendermint/blob/6291d22f46f4c4f9121375af700dbdafa51577e7/cmd/tendermint/commands/init.go#L45
-/// There exists https://github.com/informalsystems/tendermint-rs/blob/a12118978f2ffea4042d6d38ebfb290d12611314/config/src/config.rs#L23 but
-/// this seemed more straightforward as only the moniker is changed right now.
-pub fn generate_tm_config(node_name: &str, persistent_peers: &[(Id, String)]) -> String {
-    let peers_string = persistent_peers
-        .iter()
-        // https://docs.tendermint.com/master/spec/p2p/peer.html#peer-identity
-        // Tendermint peers are expected to maintain long-term persistent identities
-        // in the form of a public key. Each peer has an ID defined as
-        // peer.ID == peer.PubKey.Address(), where Address uses the scheme defined in
-        // crypto package.
-        // the peer addresses need to match this impl: https://github.com/tendermint/tendermint/blob/f2a8f5e054cf99ebe246818bb6d71f41f9a30faa/internal/p2p/address.go#L43
-        // The ID is for the node being connected to, *not* the connecting node's ID.
-        .map(|(id, ip)| format!("{}@{}", id, ip))
-        .collect::<Vec<String>>()
-        .join(",");
-    format!(
-        include_str!("../../testnets/tm_config_template.toml"),
-        node_name, peers_string,
-    )
+/// Use a hard-coded Tendermint config as a base template, substitute
+/// values via a typed interface, and rerender as TOML.
+pub fn generate_tm_config(
+    node_name: &str,
+    peers: Vec<TendermintAddress>,
+) -> anyhow::Result<String> {
+    tracing::debug!("List of TM peers: {:?}", peers);
+    let moniker: Moniker = Moniker::from_str(node_name)?;
+    let mut tm_config =
+        TendermintConfig::parse_toml(include_str!("../../testnets/tm_config_template.toml"))
+            .context("Failed to parse the TOML config template for Tendermint")?;
```

Comment:
Not worth changing here, but this could just as well be an `.expect`, since the `tm_config_template.toml` is compile-time data and so being unable to parse it is an invariant violation.